compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt

```diff
@@ -0,0 +1,385 @@
+
+        private val ANY = listOf(STRING, CHAR, BOOLEAN, BYTE, SHORT, INT, LONG, FLOAT, DOUBLE)
+
+        // Map IrConstKind to FqName
+        private val KIND_TO_FQNAME = hashMapOf<IrConstKind<*>, FqName>(
+            IrConstKind.Boolean to BOOLEAN,
+            IrConstKind.String to STRING,
+            IrConstKind.Char to CHAR,
+            IrConstKind.Byte to BYTE,
+            IrConstKind.Short to SHORT,
+            IrConstKind.Int to INT,
+            IrConstKind.Long to LONG,
+            IrConstKind.Float to FLOAT,
+            IrConstKind.Double to DOUBLE
+        )
+
+        private val UNARY_OP_TO_EVALUATOR = HashMap<UnaryOp?, Function1<Any?, Any>>()
+        private val BINARY_OP_TO_EVALUATOR = HashMap<BinaryOp?, Function2<Any?, Any?, Any>>()
+
+        @Suppress("UNCHECKED_CAST")
+        private fun <T> registerUnaryOp(operandType: FqName, operatorFN: String, f: (T) -> Any) {
+            UNARY_OP_TO_EVALUATOR[UnaryOp(operandType, operandType, operatorFN)] = f as Function1<Any?, Any>
+        }
+
+        @Suppress("UNCHECKED_CAST")
+        private fun <T> registerBuiltinUnaryOp(operandType: FqName, operatorFN: String, f: (T) -> Any) {
+            UNARY_OP_TO_EVALUATOR[UnaryOp(KOTLIN_INTERNAL_IR_FQN, operandType, operatorFN)] = f as Function1<Any?, Any>
+        }
+
+        @Suppress("UNCHECKED_CAST")
+        private fun <T1, T2> registerBinaryOp(lhsType: FqName, rhsType: FqName, operatorFN: String, f: (T1, T2) -> Any) {
+            BINARY_OP_TO_EVALUATOR[BinaryOp(lhsType, lhsType, rhsType, operatorFN)] = f as Function2<Any?, Any?, Any>
+        }
+
+        private fun <T1, T2> registerSymmetricBinaryOp(lhsType: FqName, rhsType: FqName, operatorFN: String, f: (T1, T2) -> Any) {
+            registerBinaryOp(lhsType, rhsType, operatorFN, f)
+            registerBinaryOp(rhsType, lhsType, operatorFN, f)
+        }
+
+        @Suppress("UNCHECKED_CAST")
+        private fun <T> registerBuiltinBinaryOp(operandType: FqName, operatorFN: String, f: (T, T) -> Any) {
+            BINARY_OP_TO_EVALUATOR[BinaryOp(KOTLIN_INTERNAL_IR_FQN, operandType, operandType, operatorFN)] = f as Function2<Any?, Any?, Any>
+        }
+
+        // Class Number defines toChar, toByte, ..., toString.
+        // This allows us to factor them out and save some typing.
+        private fun <T : Number> registerUnaryNumberOps(operandType: FqName) {
+            registerUnaryOp<T>(operandType, "toChar") { it.toChar() }
+            registerUnaryOp<T>(operandType, "toByte") { it.toByte() }
+            registerUnaryOp<T>(operandType, "toShort") { it.toShort() }
+            registerUnaryOp<T>(operandType, "toInt") { it.toInt() }
+            registerUnaryOp<T>(operandType, "toLong") { it.toLong() }
+            registerUnaryOp<T>(operandType, "toFloat") { it.toFloat() }
+            registerUnaryOp<T>(operandType, "toDouble") { it.toDouble() }
+            registerUnaryOp<T>(operandType, "toString") { it.toString() }
+        }
+
+        init {
```

Comment:
Is it possible to somehow use operations in `OperationsMapGenerated.kt` instead of manually enumerating them here? That would also make it future-proof in case we add more built-in operations in the future (like we did with `rem` in 1.1 for example)

Reply:
Thanks for the pointer! I updated the PR quite a bit to reuse OperationsMapGenerated.kt. Please take a look again.